docs/adr/adr1-types.md

```diff
@@ -0,0 +1,352 @@
+concrete variable names such as `a`, `b`, `c`, etc., whereas `h_1`,...,`h_k` are
+field names. The rule `T -> T` defines a function, while the rule
+`(T, ..., T) => T` defines an operator.
+
+Assuming that we have some syntax for writing down that `x` has type `T`,
+e.g., by writing `isType("x", "T")`, we can write the above examples as follows:
+
+* `x` is an integer: `isType("x", "Int")`.
+* `f` is a function from an integer to an integer: `isType("f", "Int -> Int")`.
+* `f` is a function from a set of integers to a set of integers:
+    `isType("f", "Set(Int) -> Set(Int))"`.
+* `r` is a record that has the fields `a` and `b`, where `a` is an integer
+    and `b` is a string: `isType("r", "[a |-> Int, b |-> Str])"`.
+* `f` is a set of functions from a tuple of integers to an integer:
+    `isType("f", "Set(<<Int, Int>> -> Int))"`.
+* `Foo` is an operator of an `Int` and `STRING` that returns an `Int`:
+    `isType("Foo", "(Int, Str) => Int")`.
+
+* `Bar` is a higher-order operator that takes an operator that takes
+    an `Int` and `STRING` and returns an `Int`, and returns a `BOOLEAN`:
+    `isType("Bar", "((Int, Str) => Int) => Bool")`.
+
+__Note:__ We have to pass names as strings, as it is impossible to pass operator
+    names, e.g., `Foo` and `Bar` in other operators, unless `Foo` and `Bar`
+    are nullary operators and `isType` is a higher-order operator.
+
+
+## 2. How to write type annotations (as a user)
+
+__Note__: This question is not a priority, as we do not expect the user to
+write type annotations. However, it would be good to have a solution, as sometimes
+users want to write types.
+
+Again, we have plenty of options and opinions here:
+
+1. Write type annotations by calling a special operator like `<:` or `|=`.
+1. Write type annotations as assumptions.
+1. Write type annotations in comments.
+1. Write type annotations as operator definitions.
+
+### 2.1. Type annotations with a special operator
+
+This is the current approach in Apalache. One has to define an operator, e.g., `<:`:
+
+```tla
+value <: type == value
+```
+
+Then an expression may be annotated with a type as follows:
+
+```tla
+VARIABLE S
+Init ==
+  S = {} <: {Int}
+```
+
+__Pros__:
+ * Intutive notation, similar to programming languages.
+
+__Cons__:
+
+ * This approach works well for expressions. However, it is not clear how to extend
+   it to operators.
+ * This notation is more like type clarification, rather than a type annotation.
+   Normally types are specified for names, that is, constants, variables, functions,
+   operators, etc.
+ * Same expression may be annotated in a Boolean formula. What shall we do, if the
+   user writes: `x <: BOOLEAN \/ x <: Int`?
+
+__Note:__ The current approach has an issue. If one declares the operator `<:` in
+a module `M` and then uses an unnamed instance `INSTANCE M` in a module `M2`,
+then `M` and `M2` will clash on the operator `<:`. We should define the operator
+once in a special module `Types` or `Apalache`.
```

Comment:
I actually think this would be a good idea in general, as I suspect there are likely to be useful type definitions and constructs that pop up over time.